File: interesting/powx_n.py
# https://leetcode.com/problems/powx-n/
# #recursion #log(n)
class Solution:
    
    def myPow(self, x: float, n: int) -> float:
        return self.recursiveMyPow(x, n)
        #return self.iterativeMyPow(x, n)

    def iterativeMyPow(self, x: float, n: int) -> float:
        if n == 0:
            return 1
        
        absN = n if n > 0 else -1 * n
        res = 1
        while absN > 0:
            if absN % 2 == 0:
                x *= x
                absN //= 2
            else:
                res *= x
                absN -= 1
        
        ## Another approach that I personally find more
        ## intuitive.
        # step = 1
        # while step < absN:
        #     if step * 2 <= absN:
        #         res *= res
        #         step *= 2
        #     else:
        #         res *= x
        #         step += 1

        res = res if n > 0 else 1 / res
        return res


    def recursiveMyPow(self, x: float, n: int) -> float:
        absN = n if n > 0 else -1 * n
        res = self.myPowAux(x, absN)
        res = res if n > 0 else 1/res
        return res
        
    def myPowAux(self, x: float, nonNegativeN: int):
        if nonNegativeN == 0:
            return 1
        if nonNegativeN % 2 == 0:
            res = self.myPowAux(x, nonNegativeN / 2)
            return res * res
        else:
            return x * self.myPowAux(x, nonNegativeN - 1)

File: interesting/powx_n.cc
/**
 * https://leetcode.com/problems/powx-n/
 * #recursion #log(n)
 */
class Solution {
public:
    double myPow(double x, int n) {
        ///need to cast, since if n = -2^31, 2^32 cannot be represented in
        ///int, the value of  -1 * n cannot be stored (intermidiately)
        ///thuse the coversion is done before the sign check.
        long long wideN = static_cast<long long>(n);
        unsigned long long absN = n >= 0 ? 
                static_cast<unsigned long long>(wideN) :
                static_cast<unsigned long long>(-1 * wideN);
        double res = myPowAux(x, absN);
        res = n >= 0 ? res : 1 / res;
        return res;
    }
private:
    double myPowAux(double x, unsigned long long n){
        if(n == 0){
            return 1;
        }else if(n % 2){
            return x * myPowAux(x, n - 1);
        }else{
            double res = myPowAux(x, n / 2);
            return res * res;
        }
    }
};
